lib/api/model/events.dart

```diff
@@ -439,13 +445,103 @@ class DeleteMessageEvent extends Event {
   Map<String, dynamic> toJson() => _$DeleteMessageEventToJson(this);
 }
 
-/// As in [DeleteMessageEvent.messageType].
+/// As in [DeleteMessageEvent.messageType]
+/// or [UpdateMessageFlagsMessageDetail.type].
 @JsonEnum(fieldRename: FieldRename.snake)
 enum MessageType {
   stream,
   private;
 }
 
+/// A Zulip event of type `update_message_flags`.
+///
+/// For the corresponding API docs, see subclasses.
+abstract class UpdateMessageFlagsEvent extends Event {
+  @override
+  @JsonKey(includeToJson: true)
+  String get type => 'update_message_flags';
+
+  String get op;
+
+  @JsonKey(unknownEnumValue: MessageFlag.unknown)
+  final MessageFlag flag;
+  final List<int> messages;
+  final bool all;
+
+  UpdateMessageFlagsEvent({
+    required super.id,
+    required this.flag,
+    required this.messages,
+    required this.all,
+  });
+}
+
+/// An [UpdateMessageFlagsEvent] with op `add`: https://zulip.com/api/get-events#update_message_flags-add
+@JsonSerializable(fieldRename: FieldRename.snake)
+class UpdateMessageFlagsAddEvent extends UpdateMessageFlagsEvent {
+  @override
+  String get op => 'add';
+
+  UpdateMessageFlagsAddEvent({
+    required super.id,
+    required super.flag,
+    required super.messages,
+    required super.all,
+  });
+
+  factory UpdateMessageFlagsAddEvent.fromJson(Map<String, dynamic> json) =>
+    _$UpdateMessageFlagsAddEventFromJson(json);
+
+  @override
+  Map<String, dynamic> toJson() => _$UpdateMessageFlagsAddEventToJson(this);
+}
+
+/// An [UpdateMessageFlagsEvent] with op `remove`: https://zulip.com/api/get-events#update_message_flags-remove
+@JsonSerializable(fieldRename: FieldRename.snake)
+class UpdateMessageFlagsRemoveEvent extends UpdateMessageFlagsEvent {
+  @override
+  String get op => 'remove';
+
+  final Map<int, UpdateMessageFlagsMessageDetail>? messageDetails;
+
+  UpdateMessageFlagsRemoveEvent({
+    required super.id,
+    required super.flag,
+    required super.messages,
+    required super.all,
+    required this.messageDetails,
+  });
+
+  factory UpdateMessageFlagsRemoveEvent.fromJson(Map<String, dynamic> json) =>
+    _$UpdateMessageFlagsRemoveEventFromJson(json);
+
+  @override
+  Map<String, dynamic> toJson() => _$UpdateMessageFlagsRemoveEventToJson(this);
+}
+
+/// As in [UpdateMessageFlagsRemoveEvent.messageDetails].
+@JsonSerializable(fieldRename: FieldRename.snake)
+class UpdateMessageFlagsMessageDetail {
+  final MessageType type;
+  final bool? mentioned;
+  final List<int>? userIds;
+  final int? streamId;
+  final String? topic;
```

Comment:
I think we'll probably want a crunchy shell on ensuring that when `type` is `private` we always have `userIds`, and when it's `stream` we always have `streamId` and `topic`.

That is, in zulip-flutter I think we've been pretty successful in making it so that malformed server data can't make us outright crash or misbehave, and instead just interrupts talking to the server (thanks mostly to the autogenerated `fromJson` implementations), so I'd like to try to hold the line on that as an invariant. And I suspect that doing so in the code that consumes this data, if it's just three independent nullables like this, will be annoying and make that code hard to read.

One way to get the crunchy shell would be to have a base class and two subclasses, akin to `Message` itself. But another solution might be more expedient and would also suffice: just have our `fromJson` check those conditions, and throw if they're not met, in much the same way as the autogenerated code does when an expected field is missing. For example our `fromJson` can take a few lines to do that on the `Map`, before then calling the autogenerated implementation as usual.

The downside of that approach is that the consuming code is slightly uglier, with `!` operators scattered around, than it would be with subclasses. But I think that's a good trade because there's not that much consuming code for this type, and the subclasses mean significantly more code here at the definitions. (For `Message` the tradeoff goes heavily the other way, because that's such a fundamental type and so much of our code consumes it.)

Reply:
> For example our `fromJson` can take a few lines to do that on the `Map`, before then calling the autogenerated implementation as usual.

How about calling the autogenerated implementation first? Then we can do an exhaustive switch on `type`, which will be a `MessageType` enum value.

Reply:
Sure.